tests/integ/scala/test_datatype_suite.py

```diff
@@ -472,6 +475,121 @@ def test_structured_dtypes_iceberg(
+    ):
+        pytest.skip("Test requires iceberg support and structured type support.")
+
+    read_table = f"snowpark_structured_dtypes_lqb_read_{uuid.uuid4().hex[:5]}"
+    write_table = f"snowpark_structured_dtypes_lqb_write_{uuid.uuid4().hex[:5]}"
+    query = """select
+                [1, 2, 3] :: array(bigint) as arr,
+                object_construct('k1', 1, 'k2', 2) :: map(varchar, bigint) as map,
+                1 as a,
+                2 as b
+    """
+    expected_dtypes = [
+        ("ARR", "array<bigint>"),
+        ("MAP", "map<string(16777216),bigint>"),
+        ("A", "bigint"),
+        ("B", "bigint"),
+    ]
+    expected_schema = StructType(
+        [
+            StructField("ARR", ArrayType(LongType(), structured=True), nullable=True),
+            StructField(
+                "MAP",
+                MapType(StringType(), LongType(), structured=True),
+                nullable=True,
+            ),
+            StructField("A", LongType(), nullable=True),
+            StructField("B", LongType(), nullable=True),
+        ]
+    )
+    is_query_compilation_stage_enabled = (
+        structured_type_session._query_compilation_stage_enabled
+    )
+    is_large_query_breakdown_enabled = (
+        structured_type_session._large_query_breakdown_enabled
+    )
+    original_bounds = structured_type_session._large_query_breakdown_complexity_bounds
+    try:
+        structured_type_session._query_compilation_stage_enabled = True
+        structured_type_session._large_query_breakdown_enabled = True
+        structured_type_session._large_query_breakdown_complexity_bounds = (300, 600)
+
+        create_df = structured_type_session.create_dataframe([], schema=expected_schema)
+        create_df.write.save_as_table(read_table, iceberg_config=ICEBERG_CONFIG)
+        structured_type_session.sql(
+            f"""
+        insert into {read_table}
+        {query}
+        """
+        ).collect()
+
+        base_df = structured_type_session.table(read_table)
+        assert base_df.schema == expected_schema
+        assert base_df.dtypes == expected_dtypes
+
+        df1 = base_df.with_column("A", col("A") + lit(1))
+        df2 = base_df.with_column("B", col("B") + lit(1))
+
+        for i in range(6):
+            df1 = df1.with_column("A", col("A") + lit(i) + col("A"))
+            df2 = df2.with_column("B", col("B") + lit(i) + col("B"))
+
+        df1 = df1.group_by(col("A")).agg(
+            sum_distinct(col("B")).alias("B"),
+            any_value(col("ARR")).alias("ARR"),
+            any_value(col("MAP")).alias("MAP"),
+        )
+        df2 = df2.group_by(col("B")).agg(
+            sum_distinct(col("A")).alias("A"),
+            any_value(col("ARR")).alias("ARR"),
+            any_value(col("MAP")).alias("MAP"),
+        )
+        union_df = df1.union_all(df2)
+        union_df = union_df.select(
+            array_sort("ARR", sort_ascending=False).alias("ARR"), "MAP", "A", "B"
+        )
```

Comment:
can we do a check of union_df.schema here to see if the schema is the same before and after